pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -284,31 +283,15 @@ func (c *Reconciler) reconcile(ctx context.Context, pr *v1beta1.PipelineRun) err
 		pr.ObjectMeta.Annotations[key] = value
 	}
 
-	d, err := dag.Build(v1beta1.PipelineTaskList(pipelineSpec.Tasks))
-	if err != nil {
-		// This Run has failed, so we need to mark it as failed and stop reconciling it
-		pr.Status.MarkFailed(ReasonInvalidGraph,
-			"PipelineRun %s/%s's Pipeline DAG is invalid: %s",
-			pr.Namespace, pr.Name, err)
-		return controller.NewPermanentError(err)
-	}
-
-	// build DAG with a list of final tasks, this DAG is used later to identify
-	// if a task in PipelineRunState is final task or not
-	// the finally section is optional and might not exist
-	// dfinally holds an empty Graph in the absence of finally clause
-	dfinally, err := dag.Build(v1beta1.PipelineTaskList(pipelineSpec.Finally))
-	if err != nil {
-		// This Run has failed, so we need to mark it as failed and stop reconciling it
-		pr.Status.MarkFailed(ReasonInvalidGraph,
-			"PipelineRun %s's Pipeline DAG is invalid for finally clause: %s",
-			pr.Namespace, pr.Name, err)
-		return controller.NewPermanentError(err)
-	}
-
 	if err := pipelineSpec.Validate(ctx); err != nil {
 		// This Run has failed, so we need to mark it as failed and stop reconciling it
-		pr.Status.MarkFailed(ReasonFailedValidation,
+		var reason = ReasonFailedValidation
+		if err.Details == "Invalid Graph" {
```

Comment:
Because the graph is created in `pipelinerunresources`, validation happens first. In this context it was not clear that the error was an invalid graph one, so I added some extra context in the error to be able to catch it and use the specific reason. 
The alternatives are:
- skip validation here, but that's only a temp solution
- move validation into `ResolvePipelineRun` - which might make sense, but not in this PR
- simply return a validation error and rely on the error stack to provide more insight to the user
I wonder if we really need to run validation here, but I guess we will need to do once tasks / pipelines may come from a container registry or git url.

Reply:
@afrittoli `pipelineSpec` is being validated here for many different checks in addition to `invalidGraph`. I think we do need to run validation here, I dont see validation happening else where, am I missing something? 🤔

Reply:
We could certainly move the check on invalid graph to `ResolvePipelineRun`. 

Also, such reason could be generated where the error is getting created in `pipeline_validation.go` instead of calculating here. Its a great idea to set the reason to specific validation error in addition to returning `error`.

Reply:
> @afrittoli `pipelineSpec` is being validated here for many different checks/conditions in addition to `invalidGraph`. I think we do need to run validation here, I dont see validation happening else where, am I missing something? 🤔

Validation is done by the webhook when resources are submitted to etcd.
However it will not be performed for resources that are not stored in etcd like tasks or pipelines coming from an external source (git/registry). I'm not sure about embedded specs - I'd expect to go through the same validation as normal resources.

Reply:
> We could certainly move the check on invalid graph to `ResolvePipelineRun`.

That's the way I implemented it originally, but because we validate the spec here, we don't really catch the error in `ResolvePipelineRun`. The only way would be to move the whole validation of the spec into `ResolvePipelineRun`, which is something we could do eventually, but I didn't want to do it here as well.

> 
> Also, such reason could be generated where the error is getting created in `pipeline_validation.go` instead of calculating here. Its a great idea to set the reason to specific validation error in addition to returning `error`.